neps/nep-0508.md

```diff
@@ -0,0 +1,231 @@
+
+In this NEP we propose that resharding should be rolled out first, before any real dependency on state sync is added. We can then safely roll out the resharding logic and solve the above mentioned issues separately. We believe atleast some of the issues can be mitigated by the implementation of new pre-state root and chunk execution design.
+
+## Integration with Stateless Validation
+
+The Stateless Validation requires that chunk producers provide proof of correctness of the transition function from one state root to another. That proof for the first block after the new shard layout takes place will need to prove that the entire state split was correct as well as the state transition.
+
+In this NEP we propose that resharding should be rolled out first, before stateless validation. We can then safely roll out the resharding logic and solve the above mentioned issues separately.
+
+## Future fast-followups
+
+### Resharding should work even when validators stop tracking all shards
+
+As mentioned above under 'Integration with State Sync' section, initial release of resharding v2 will happen before the full implementation of state sync and we plan to tackle the integration between resharding and state sync after the next shard split (Won't need a separate NEP as the integration does not require protocol change.)
+
+### Resharding should work after stateless validation is enabled
+
+As mentioned above under 'Integration with Statelss Validation' section, initial release of resharding v2 will happen before the full implementation of stateless validation and we plan to tackle the integration between resharding and stateless validation after the next shard split (May need a separate NEP depending on implemetnation detail.)
+
+## Future possibilities
+
+### Dynamic resharding
+
+As noted above, dynamic resharding is out of scope for this NEP and should be implemented in the future. Dynamic resharding includes the following but not limited to:
+
+* Automatic determination of split boundary based on parameters like traffic, gas usage, state size, etc.
+* Automatic shard splitting and merging
+
+### Localization of resharding event to specific shard
+
+As of today, at the RocksDB storage layer, we have the ShardUId, i.e. the ShardId along with the ShardVersion, as a prefix in the key of trie state and flat state. During a resharding event, we increment the ShardVersion by one, and effectively remap all the current parent shards to new child shards. This implies we can't use the same underlying key value pairs for store and instead would need to duplicate the values with the new ShardUId prefix, even if a shard is unaffected and not split.
+
+In the future, we would like to potentially change the schema in a way such that only the shard that is splitting is impacted by a resharding event, so as to avoid additonal work done by nodes tracking other shards.
+
+### Other useful features
+
+* Account colocation for low latency across account call
```

Comment:
I've never heard about this being considered - would it be useful to elaborate a little more?

Also for "Shard on demand" what does that mean?

Reply:
* colocation - I think this relates to an idea that was dicussed on the side during the Mardid offsite but I'm not 100% sure about the details. 
* shard on demand - I'm not certain what exactly does that mean. 

@walnut-the-cat did you add those? Can you clarify it a bit?